src/approval/index.ts

```diff
@@ -0,0 +1,144 @@
+
+  const cfnTemplate = await fs.readFileSync(stackArgs.Template);
+  const cfnTemplateFileName = stackArgs.Template.match(/[a-zA-Z-_0-9]*\.yaml/)![0];
+  const s3Url = parseS3HttpUrl(stackArgs.ApprovedTemplateLocation);
+
+  let hashedKey = new Md5().appendStr(cfnTemplate.toString()).end().toString();
+
+  if (s3Url.key.length > 0) {
+    hashedKey = `${s3Url.key}/${hashedKey}`;
+  }
+
+  try {
+    await s3.headObject({
+      Bucket: s3Url.bucket,
+      Key: `${hashedKey}.yaml`
+    }).promise();
+
+    logSuccess(`👍 Your template has already been approved`);
+  } catch (e) {
+    if (e.code === "NotFound") {
+      await s3.putObject({
+        Body: cfnTemplate,
+        Bucket: s3Url.bucket,
+        Key: `${hashedKey}.yaml.pending`
+      }).promise();
+
+      logSuccess(`Successfully uploaded ${cfnTemplateFileName} to ${stackArgs.ApprovedTemplateLocation}`);
+      logSuccess(`Approve template with: `);
+      console.log(`  iidy approve-template https://s3.amazonaws.com/${s3Url.bucket}/${hashedKey}.yaml.pending`);
+    } else {
+      throw new Error(e);
+    }
+  }
+
+  return 0;
+}
+
+export async function approveTemplate(argv: Arguments): Promise<number> {
+  await configureAWS(argv.profile, "us-east-1");
+  const s3 = new S3();
+  const s3Url = parseS3HttpUrl(argv.filename);
+  const bucketDir = path.parse(s3Url.key).dir;
+
+  try {
+    await s3.headObject({
+      Bucket: s3Url.bucket,
+      Key: s3Url.key.replace(/\.pending$/, "")
+    }).promise();
+
+    logSuccess(`👍 The template has already been approved`);
+  } catch (e) {
+    if (e.code === "NotFound") {
+
+      const pendingTemplate = await s3.getObject({
+        Bucket: s3Url.bucket,
+        Key: s3Url.key
+      }).promise().then((template) => template.Body );
+
+      // When `--approve true` is not supplied, it will diff the file for review
+      // command has to be run again with the flag to approve changes
+      if (argv.approve === false) {
+        const previouslyApprovedTemplate = await s3.getObject({
+          Bucket: s3Url.bucket,
+          Key: `${bucketDir}/latest.yaml`
+        }).promise()
+          .then((template) => template.Body)
+          .catch((e) => {
+            if (e.code !== "NoSuchKey") {
+              return Promise.reject(e);
+            }
+            return Buffer.from("")
+          });
+
+        const diff = jsdiff.diffLines(
+          pendingTemplate!.toString(),
+          previouslyApprovedTemplate!.toString());
+        let colorizedString = ""
+
+        diff.forEach(function(part){
+          if (part.added) {
+            colorizedString = colorizedString + cli.green(part.value)
+          } else if (part.removed) {
+            colorizedString = colorizedString + cli.red(part.value)
+          }
+        });
+        console.log(colorizedString);
+        console.log("Do these changes look good for you? Add a flag `--approve true` to approve");
+
+        return 0
+      }
+
+      // create a new pending file
```

Comment:
This should probably be in the `else` branch of the conditional above (even though there is a `return`).